### PR TITLE
Discover crate type information late if necessary.

### PR DESCRIPTION
**Pull request by Mark-Simulacrum**
Some crates aren't found during eager crate-type searching due to being behind `cfg(...)` flags. We still want to be able to handle these, though, so when necessary we now call rustc again to get crate-type information for these cfg-ed crates.

This is necessary for https://github.com/rust-lang/rust/pull/41991.

r? @alexcrichton

### CODE REVIEW

**Comment by Mark-Simulacrum** [comment]
Updated, I think this fixes the tests.

**Comment by alexcrichton** [comment]
Thanks! This I think only has a downside of running rustc more times, right? No-op builds of Cargo can sometimes be dominated by this as we've gotta run rustc *three* times now which actually blows up to 6 processes if you're running rustup. Clocking in at a minimum of 5ms per process (ish) we're getting close to 30-50ms overhead in Cargo for "doing nothing", which can get pretty pricy.

Can you describe the original bug in some more detail? Would it be possible to not need to spawn a new rustc process? I've long intended to have a fancier caching mechanism for rustc processes where we only rarely spawn rustc to learn this sort of info, but that's still a long ways off unfortunately.

**Comment by Mark-Simulacrum** [comment]
I think it's possible we can spawn less rustc processes here (in theory we'd only need to do this cfg/sysroot discovery twice, once for `Kind::Target` and once for `Kind::Host`. That might already be what we do, I'm not sure how often the relevant function is called.

Basically, the initial bug here was that we wouldn't look at dependencies which had `cfg(..)` in their definition in Cargo.toml when attempting to find out what crate types a user had. The problem is likely possible to notice not only with proc-macro crates (most unusual crate types will do), but it's easiest to notice with them. I'm not sure how much we can do here to mitigate or remove the problem entirely; I can't think of anything myself. Let me know if that helps clarify things for you.

**Comment by alexcrichton** [comment]
Hm ok, thanks! I wonder if we could traverse the whole crate graph, regardless of `cfg` on the dependencies, to collect the crate types we should be learning about?

**Comment by Mark-Simulacrum** [comment]
Hm, maybe. Although it would mean we'd traverse some crates that we *shouldn't* (i.e., doing extra work) but that does seem plausible. I'll try to make that change here as soon as I get a chance.

**Comment by alexcrichton** [comment]
The "extra work" here is fine, I don't think it'll cause any problems.

**Comment by Mark-Simulacrum** [comment]
Okay, pushed up a new version. Note that a few tests are changed/deleted since they either no longer make sense or need to be updated for the new state of things. I'd appreciate comments if there's some area we can do better here to decrease what we have to download.

File: tests/cfg.rs [hunk 1]
```diff
@@ -247,7 +247,8 @@ fn ignore_version_from_other_platform() {
     assert_that(p.cargo_process("build"),
                 execs().with_status(0).with_stderr("\
 [UPDATING] registry [..]
-[DOWNLOADING] [..]
+[DOWNLOADING] foo [..]
+[DOWNLOADING] foo [..]
```
Comment (alexcrichton) [inline tests/cfg.rs]: Argh I forgot about this! This is actually a pretty crucial feature today where builds on Linux don't download Windows-only dependencies.

**Comment by Mark-Simulacrum** [comment]
So we can revert to the patch which runs rustc more times, or keep this and just live with having to download dependencies that we won't actually use. As noted in the commit, it's possible we can change the crates.io API to expose crate-type more directly (without download of the whole crate).

So an alternative approach here is to not query rustc for the prefix/suffix for each crate type, but instead copy that code into Cargo. I assume it's something like a table, though I don't know, in which case maintaining it should be fairly easy. Obviously, it does add some level of overhead to adding new crate types, but adding new crate types is rare enough that it may be worth it. It's not really that much information, though if it varies depending on target platform or other details that's a negative point towards duplicating the information.

Maybe we can alternatively always execute rustc with `-o` and output into a Cargo-specific file? For example, we'd build rlibs into `rlib-cratename` or `cratename.rlib` or something like that. I guess that'd probably confuse rustc and linkers, though.

-----

At this point, it may be worth considering that I'm not sure that the cost of downloading windows dependencies is all that severe, though it may be higher than a slight overhead on all Cargo invocations of calling rustc more times (as the earlier version of the patch did).

I think that this is a bug that needs to be fixed, one way or another, or at least a workaround proposed. It's hitting actual users and is painful to work around. I'm not sure the overhead you [noted](https://github.com/rust-lang/cargo/pull/4358#issuecomment-320416599) (30-50 ms) is severe enough to justify not going with that approach because of it. Arguably, though, the current implementation has less overhead in the case of doing nothing, since we've already downloaded all dependencies.

**Comment by alexcrichton** [comment]
Hm well so just to make sure I've got a handle on this, is this really that much of a widespread problem? I'm under the impression that it's basically limited to "target specific dependencies" which are actually misused for `--cfg` flags passed through `RUSTFLAGS`, basically rustc's and only rust's use case. AFAIK this has never come up in the wild in Cargo.

Despite that though I'm relatively firm on (a) not adding more process invocations and (b) not downloading extraneous dependencies. More invocations can seriously affect noop build time and more downloads skews statistics and pollutes caches. Basically budging on either of this is *guaranteed* to mean a future bug report that we'll need to fix.

It's a good though though to even question why Cargo is doing this! Certainly for a bunch of targets (e.g. almost all built in ones) Cargo can statically know this information without asking the compiler itself. The catch is custom targets, however, which can arbitrarily change their filenames. In that sense it's impossible I think for this logic to 100% live in Cargo without reimplementing custom target specifications from the compiler.

What we *may* be able to do is whitelist some targets and hardcode knowledge, but it doesn't really solve the problem. Another possible solution though is to dynamically detect this information in these pessimistic cases (which I'd hypothesize are not common). That is, if we don't have cached information for what a crate name will be we immediately run the compiler to learn such info. Most Cargo compilations won't have this extra compile, but some will.

**Comment by Mark-Simulacrum** [comment]
> I'm under the impression that it's basically limited to "target specific dependencies" which are actually misused for --cfg flags passed through RUSTFLAGS, basically rustc's and only rust's use case. AFAIK this has never come up in the wild in Cargo.

I believe that this hasn't yet come up in the wild yet because no one has tried to depend on a crate with a non-whitelisted crate-type behind `cfg` of any kind (e.g., target platform). This is likely because I'd guess most Cargo projects out there are rlibs, some are dylibs, and a few are proc-macro crates. proc-macro crates in general aren't target specific, though I could imagine something like bindgen wanting to target only a specific platform. It does seem possible to encounter though; all you'd need is a crate like winapi to depend on serde, I think.

> Another possible solution though is to dynamically detect this information in these pessimistic cases (which I'd hypothesize are not common)

I think this seems like a viable solution to the problems this PR faces; I'll try to implement it soon.

**Comment by Mark-Simulacrum** [comment]
@alexcrichton Updated with the late-discovery method. Generally seems to work quite well; no tests failed locally.

File: src/cargo/ops/cargo_rustc/context.rs [hunk 2]
```diff
@@ -564,6 +566,47 @@ impl<'a, 'cfg> Context<'a, 'cfg> {
...
+        };
+        let mut parts = line.trim().split("___");
+        let prefix = parts.next().unwrap();
+        let suffix = match parts.next() {
+            Some(part) => part,
+            None => bail!("output of --print=file-names has changed in \
+                the compiler, cannot parse"),
+        };
+
+        Ok(Some((prefix.to_string(), suffix.to_string())))
```
Comment (Mark-Simulacrum) [inline src/cargo/ops/cargo_rustc/context.rs]: Most of this function is copied from `probe_target_info_kind` but I don't think we can do much about that -- I don't see an easy way to deduplicate much of it in a clean way. Open to suggestions though.

**Comment by alexcrichton** [comment]
Thanks yeah looks good! I think right now though we need to remove the duplication because there's features like `RUSTFLAGS` handling which aren't shared between these two code paths. Perhaps the original function could have a structured return value and the new invocation being added would just handle a few fields coming out of that? It should be essentially "free" to call one rustc invocation to learn about extra information and then discard it.

**Comment by Mark-Simulacrum** [comment]
Hm, I'm not entirely sure I follow you. You mean we'd stash the rustflags and other arguments into a struct to reuse later?

**Comment by alexcrichton** [comment]
Er sorry, basically we need to share the definitions here, I'd prefer to not land two copies of the same function. However that happens I don't mind.

**Comment by Mark-Simulacrum** [comment]
Alright, I think I sufficiently deduplicated the two code paths. Let me know what you think.

**Comment by alexcrichton** [comment]
@bors: r+

Thanks!

**Pull request merged**